### ISSUE TITLE
Tpetra:  Nalu regression testing fails due to recent Trilinos changes

### ISSUE THREAD

**Issue by wjhorne** [trilinos/Trilinos#10463]
## Bug Report
@trilinos/tpetra

### Description
Nalu's  (@spdomin) nightly regression test suite currently fails due to recent changes to Trilinos. A git bisect yields the output below. From what we can tell there are new, extremely small diffs in residuals due to the changes in the commit below. This could be due to something as small as an order of operations change or similar. Could we get some verification that the changes are correct so that we can re-bless our tests accordingly?  

72314b8136aa3980e339d186be6d4bc8242c7506 is the first bad commit
commit 72314b8136aa3980e339d186be6d4bc8242c7506
Author: Christian Glusa <caglusa@sandia.gov>
Date:   Wed Apr 6 12:43:21 2022 -0600

    Tpetra: Use numVectors==1 specialization for local parts of norms

:040000 040000 475b4cef23ffb399e71a1e3749d5ae193b1a5ba1 d9be457f73757da6aa0ee8d1480b96750f5b0dc2 M packages

### Steps to Reproduce
1. Checkout commit shown above 
2. Link resulting build to latest Nalu build

**Comment by cgcgcg** [comment]
@trilinos/tpetra

**Comment by cgcgcg** [comment]
We switched from using a multivector kernel of the norm to a single vector kernel when a multivector only has one vector. I'm not sure why this would give a slightly different result. @brian-kelley

**Comment by brian-kelley** [comment]
@cgcgcg It would change the associativity of the additions. The single vector kernel does whatever Kokkos::parallel_reduce does, but the multivector kernel breaks each column into blocks and atomically adds the partial sums into the result. 

However, I'm surprised this would cause a Nalu diff now because before I merged in #10028 on Jan. 13, which accidentally took out this single-vector special case from the multivector kernel in KokkosKernels. But that didn't cause a regression, as far as I know.

So basically the behavior of Tpetra's norms with 1 column was:

pre-Jan 13: parallel_reduce
Jan 13 - Apr 6: multivector kernel, blocks combined with atomic_add (slow for 1 column)
Apr 6 - now: parallel_reduce again

BTW, since Kokkos 3.6 was merged today, I'm working on a permanent PR to restore the numVecs=1 special path in Kokkos Kernels. This way Tpetra doesn't need to have the special path. But it won't have an effect on this regression.

**Comment by cgcgcg** [comment]
@brian-kelley Could this have been the issue in January? #10086

**Comment by jhux2** [comment]
@wjhorne How large are the residual diffs in Nalu?

**Comment by spdomin** [comment]
@jhux2, they are very small - 1.0e-6ish. However, in the past such small diffs pointed to real (undesirable) changes.

I thought #10028 did cause diffs?

https://github.com/trilinos/Trilinos/issues/10086

At any rate, all we need is a thumbs up that these are okay to accept.

Best,

**Comment by brian-kelley** [comment]
Oh yeah... I did get mentioned once on the #10086 issue but I probably missed the notification. I just opened #10465 which will permanently fix the performance regression. For this issue, I would say re-bless the output you get with Christian's PR now, and then there shouldn't be another regression when this new PR gets merged.

**Comment by spdomin** [comment]
@wjhorne, please do re-bless the results using the SHA1 of associated with the first diff. Then, move to the most recent Trilinos/develop; test that version to ensure that no diffs maneuvered in during that three day time period. If all is well, issue a PR on Nalu and let us know here that all is well to close.

**Comment by jhux2** [comment]
@jrood-nrel

**Comment by spdomin** [comment]
A FYI, looks like the tests this morning failed. That may complicate the issue. @wjhorne, let us know.

**Comment by wjhorne** [comment]
@spdomin I will have the final results of the re-bless and comparison soon and will update here. There might be a diff in-between that I am investigating, but I am not certain quite yet.

**Comment by wjhorne** [comment]
I re-blessed using the first commit mentioned above and then looked at the diffs to the most recent Trilinos develop build. All tests are passing with both Trilinos builds after re-blessing. I had to use the cmake set policy fix mentioned elsewhere to install the most recent build. 

I can put together a PR that includes the cmake policy change and the re-bless assuming everything else looks good to others here.

**Comment by spdomin** [comment]
Please do. Also, please update the Trilinos config file. Thanks.

**Comment by spdomin** [comment]
All is well on our side. We are diff-free and fully configured. 

Best
